dataikuapi/dss/ml.py

```diff
@@ -308,50 +332,679 @@ def set_metric(self, metric=None, custom_metric=None, custom_metric_greater_is_b
+        self._raw_settings["strategy"] = strategy
+
+    def set_validation_mode(self, mode):
+        assert mode in {"SHUFFLE", "KFOLD", "TIME_SERIES_SINGLE_SPLIT", "TIME_SERIES_KFOLD", "CUSTOM"}
+        self._raw_settings["mode"] = mode
+
+    def set_validation_stratified(self, stratified):
+        assert isinstance(stratified, bool)
+        self._raw_settings["stratified"] = stratified
+
+    def set_validation_n_folds(self, n_folds):
+        assert isinstance(n_folds, int)
+        if self._raw_settings["mode"] not in {"KFOLD", "TIME_SERIES_KFOLD"}:
+            warnings.warn("\'nFolds\' parameter will be ignored in {} mode".format(self._raw_settings["mode"]))
+        self._raw_settings["nFolds"] = n_folds
+
+    def set_validation_split_ratio(self, split_ratio):
+        assert isinstance(split_ratio, float)
+        if self._raw_settings["mode"] not in {"SHUFFLE", "TIME_SERIES_SINGLE_SPLIT"}:
+            warnings.warn("\'splitRatio\' parameter will be ignored in {} mode".format(self._raw_settings["mode"]))
+        self._raw_settings["splitRatio"] = split_ratio
+
+    def set_validation_n_shuffle_iter(self, n_shuffle_iter):
+        assert isinstance(n_shuffle_iter, int)
+        if self._raw_settings["mode"] != "SHUFFLE":
+            warnings.warn("\'shuffleIterations\' parameter will be ignored in {} mode".format(self._raw_settings["mode"]))
+        self._raw_settings["shuffleIterations"] = n_shuffle_iter
+
+    def set_execution_timeout(self, timeout):
+        assert isinstance(timeout, int)
+        self._raw_settings["timeout"] = timeout
+
+    def set_execution_n_iter(self, n_iter):
+        assert isinstance(n_iter, int)
+        if self._raw_settings["strategy"] == "GRID":
+            self._raw_settings["nIter"] = n_iter
+        else:
+            self._raw_settings["nIterRandom"] = n_iter
+
+    def set_grid_search_shuffling(self, shuffled, seed=None):
+        assert isinstance(shuffled, bool)
+        if seed is not None:
+            assert isinstance(seed, int)
+            self._raw_settings["seed"] = seed
+        self._raw_settings["randomized"] = shuffled
+        if self._raw_settings["strategy"] != "GRID":
+            warnings.warn("\'randomized\' and \'seed\' parameters will be ignored in {} strategy".format(self._raw_settings["strategy"]))
+
+    def set_search_distribution(self, distributed, n_containers):
+        assert isinstance(distributed, bool)
+        if n_containers is not None:
+            assert isinstance(n_containers, int)
+            self._raw_settings["nContainers"] = n_containers
+        self._raw_settings["distributed"] = distributed
+
+    def set_n_jobs(self, n_jobs):
+        assert isinstance(n_jobs, int) and n_jobs > 0
+        self._raw_settings["nJobs"] = n_jobs
+
+
+class HyperparameterSettings(object):
+
+    def __init__(self, name, algo_settings):
+        self.name = name
+        self._algo_settings = algo_settings
```

Comment:
Not sure about that second proposal, seems more complex.

Reply:
Not sure I get it, `AlgorithmSettings` _is_ the hyperparameter settings? It allows to keep the reference (not copy of the whole thing, just a reference) on the rest so that you can filter out irrelevant (inactive) parts in the repr.

Extending `dict` is for backward compatibility, not sure we want to do it more than necessary unless we have a good reason to do so.

Reply:
I agree that the current code only keeps the reference (not the copy) of `AlgorithmSettings`. 

It's just a matter of design, since I was intuitively thinking about `HyperparameterSettings` as being part of the `AlgorithmSettings`, while the current state is that all the info is in `AlgorithmSettings`, and `HyperparameterSettings` only provide the methods to correctly print / modify the relevant parts in the `AlgorithmSettings` object.

In the end, the API usage would remain the same,

Reply:
@mhham I think the original reason was that I needed the `strategy` to define the actual `mode` (either `gridMode` or `randomMode`) in the `HyperparameterSettings`, so I just kept the pointer to the larger object with the name of the hyperparameter field.